src/app/shared/services/chat/offline/chat.flow.ts

```diff
@@ -0,0 +1,207 @@
+import { Observable, of, BehaviorSubject } from 'rxjs';
+import { ChatMessage, ChatResponseOption } from '../chat-msg.model';
+import { FlowStatusChange } from './offline-chat.service';
+import { RapidProFlowExport } from './rapid-pro-export.model';
+
+export interface ChatFlow {
+    sendMessage(msg: ChatMessage): Observable<any>;
+    messages$: BehaviorSubject<ChatMessage[]>;
+    flowStatus$: BehaviorSubject<FlowStatusChange[]>;
+}
+
+export class RapidProOfflineFlow implements ChatFlow {
+
+    name: string;
+    nodesById: { [nodeUUID: string]: RapidProFlowExport.Node } = {};
+    currentNode: RapidProFlowExport.Node;
+    childFlowId: string = null;
+
+    constructor(protected flowObject: RapidProFlowExport.Flow, public messages$: BehaviorSubject<ChatMessage[]>,
+        public flowStatus$: BehaviorSubject<FlowStatusChange[]>, public contactFields: { [field: string]: string }) {
+        console.log("Export object!", flowObject);
+        this.name = flowObject.name;
+        this.flowObject.nodes.forEach((node) => {
+            this.nodesById[node.uuid] = node;
+        });
+        this.flowStatus$.subscribe((flowStatusChanges) => {
+            if (flowStatusChanges.length > 0) {
+                let latest = flowStatusChanges[flowStatusChanges.length - 1];
+                if (this.flowObject.uuid === latest.flowId && latest.status === "start") {
+                    console.log("Entered node by status change sub");
+                    this.enterNode(this.flowObject.nodes[0]);
+                }
+            }
+        });
+    }
+
+    private enterNode(node: RapidProFlowExport.Node) {
+        this.currentNode = node;
+        console.log("Entered node ", node.uuid, node);
+        for (let action of node.actions) {
+            if (action.type === "enter_flow") {
```

Comment:
This might be better as a switch statement so that we can include a default case to at least log/report if there are any cases not handled. Also would reduce the number of if statements which is a bit confusing (but I'm assuming required given a lot of possible cases... although you could sacrifice a bit of efficiency for readability and just calculate things like latestFlowEvent at a higher level regardless of action type). Similarly you could add getter functions for things like latestFlow that could just return if exists, e.g.  `get latestFlow(){return flowEvents.length > 0 ? flowEvents[flowEvents.length-1] : null`  and call directly `if(latestFlow){...}